### DISCUSSION TITLE
How to know if there's a corresponding/counterpart video for the song?

### DISCUSSION THREAD

**Discussion by alex29ua** [sigma67/ytmusicapi#390] [Q&A]
I'm willing to get a list of official videos for my music library. Those that appear in the "Song/Video" tab switch sometimes. 
So far no luck. I have even tried debugging the browser page to find when the "Video" switch becomes active, and how it knows the video ID, but it doesn't look openly available in the responses. 

I got a list of all my songs in the library (just "Google download / takeout" feature), so I have their IDs. 
The thing that might work, but kind of ugly and not fully tested, is to get the album id for each song, and then use the `get_album()`  call - and only then the `MUSIC_VIDEO_TYPE_OMV` value appears. Otherwise, when I do `get_song()` call, there's no indication of video being available. Also those "MUSIC_VIDEO_TYPE_" seem to be not what I'd expect. 

What I have also noticed is that the IDs produced by `get_album()` are different from what the YTmusic website uses, and they seem to be the ids of the actual corresponding videos. 

For example, what I use for testing: there's a song `-vIAjs5nGFo` and the album id for it is `MPREb_8XwWknRIRMO`. And when I do `get_album()`, I get a different ID for this song, `MtHEiuuYt8k` (which happens to be the actual video id as well).

Any suggestions about this confusion and about the easiest way of finding the videos for songs?

**Comment by sigma67** [comment]
Counterpart can be identified via get_watch_playlist: https://ytmusicapi.readthedocs.io/en/stable/reference.html#ytmusicapi.YTMusic.get_watch_playlist

BrowseId vs VideoId (the MPRE.. is a browseId): https://ytmusicapi.readthedocs.io/en/stable/faq.html#what-is-a-browseid

In general, use an authenticated instance if you want to get only songs.

Does this help?

**Comment by polvallverdu** [comment]
It's true that using a unauthorized user returns `MUSIC_VIDEO_TYPE_OMV`s for albums, but what about playlists. Playlists with actual songs, unauthorized users also get `MUSIC_VIDEO_TYPE_ATV`. Is there any other way to know the id of the counterpart?

**Reply by ImpenetrableNoble** [reply]
@polvallverdu For every song in your library, use `the get_album()` method to retrieve the album details. This will give you information about the song, including any official music videos.

- Example:
```
from ytmusicapi import YTMusic

ytmusic = YTMusic()
album_id = "MPREb_8XwWknRIRMO"
album_details = ytmusic.get_album(album_id)

for track in album_details['tracks']:
    if 'videoId' in track:
        print(f"Song ID: {track['videoId']}, Video ID: {track['videoId']}")
```

Within the album details, check for a videoId field. This field will contain the ID of the official music video if it exists.

Example:
```
song_id = "vIAjs5nGFo"
album_id = "MPREb_8XwWknRIRMO"

album_details = ytmusic.get_album(album_id)
watch_playlist = ytmusic.get_watch_playlist(song_id)

video_ids = set()

for track in album_details['tracks']:
    if 'videoId' in track:
        video_ids.add(track['videoId'])

if 'contents' in watch_playlist:
    for item in watch_playlist['contents']:
        if 'videoId' in item:
            video_ids.add(item['videoId'])

print(f"Video IDs for Song ID {song_id}: {video_ids}")


```
Use both methods to cross-verify and ensure you get the correct video ID.

For example:
```
song_id = "vIAjs5nGFo"
album_id = "MPREb_8XwWknRIRMO"

album_details = ytmusic.get_album(album_id)
watch_playlist = ytmusic.get_watch_playlist(song_id)

video_ids = set()

for track in album_details['tracks']:
    if 'videoId' in track:
        video_ids.add(track['videoId'])

if 'contents' in watch_playlist:
    for item in watch_playlist['contents']:
        if 'videoId' in item:
            video_ids.add(item['videoId'])

print(f"Video IDs for Song ID {song_id}: {video_ids}")

```
By combining the `get_album()` and `get_watch_playlist()` methods, you can identify the video IDs for your music library.

It'll leverage the strength of both methods to ensure you capture the correct video IDs, even without relying solely on authentication.

And if you want to, you can Implement error handling to manage situations where a video ID is not found or an API call fails to make sure 
that your process is robust and reliable.
